File: src/mating.py
"""
mating.py
=========
Decide which faces can be *glued face-to-face* on an identical copy, and compute
the rigid transform that performs the join.

Why this is non-trivial
-----------------------
Two faces being congruent is **not** enough to mate them. For a flush joint the
two faces meet with their outward normals opposed, which means (viewed from one
side) one polygon must coincide with the *mirror image* of the other. So:

* If face A and face A' are related by a **reflection** (opposite chirality) or A
  is self-mirror-symmetric, they CAN mate.
* If they are related only by a **rotation** and are chiral, they CANNOT mate --
  the mirror-shaped face they would need does not exist on an identical copy.

We detect this numerically: search reversed cyclic vertex correspondences (the
orientation flips because the normals oppose), run a *proper* (det=+1) Kabsch
fit, and accept a mate only when RMSD ~ 0 AND the normals end up anti-parallel.
"""

from __future__ import annotations
import numpy as np
from numpy.linalg import svd, det

from geometry import Face


def kabsch(A: np.ndarray, B: np.ndarray):
    """Best PROPER rigid motion (rotation R det=+1, translation t) mapping the
    ordered point set A onto B. Returns (R, t, rmsd)."""
    ca, cb = A.mean(0), B.mean(0)
    H = (A - ca).T @ (B - cb)
    U, S, Vt = svd(H)
    d = np.sign(det(Vt.T @ U.T))
    R = Vt.T @ np.diag([1, 1, d]) @ U.T
    t = cb - R @ ca
    rmsd = float((((R @ A.T).T + t - B) ** 2).sum(1).mean() ** 0.5)
    return R, t, rmsd


def mate_transform(face_j: Face, face_i: Face):
    """Proper rigid motion (R, t) that lays ``face_j`` flush onto ``face_i``
    (centroids coincide, normals opposed, outlines aligned).

    Returns (R, t, rmsd, normal_dot). A clean mate has rmsd ~ 0 and
    normal_dot ~ -1. Returns None if the faces have different vertex counts.
    """
    A0 = face_j.pts[::-1]                       # reverse: opposing normals flip orientation
    B = face_i.pts
    if len(A0) != len(B):
        return None
    best = None
    for s in range(len(B)):
        R, t, r = kabsch(np.roll(A0, s, axis=0), B)
        if best is None or r < best[2]:
            nd = float((R @ face_j.normal) @ face_i.normal)
            best = (R, t, r, nd)
    return best


def flush_mate_graph(faces: list[Face], rmsd_tol: float = 1e-2):
    """For every face, the list of faces it can be glued flush against.

    Returns dict ``{face_idx: [mateable_face_idx, ...]}``. A face whose list is
    empty cannot participate in a magnetic face joint on an identical copy.
    """
    by_idx = {f.idx: f for f in faces}
    graph: dict[int, list[int]] = {}
    for i in faces:
        mates = []
        for j in faces:
            res = mate_transform(by_idx[j.idx], by_idx[i.idx])
            if res is None:
                continue
            _, _, rmsd, nd = res
            if rmsd < rmsd_tol and nd < -0.99:
                mates.append(j.idx)
        graph[i.idx] = sorted(mates)
    return graph


def derive_connections(graph: dict[int, list[int]]):
    """Turn the mate graph into a list of connections and the unusable faces.

    * faces with an empty mate list  -> ``unusable``
    * a face that only mates with itself -> self-connection ``(i, i)``
    * a congruent mateable pair {a, b}   -> connection ``(a, b)``
    * larger mateable components are chained pairwise (with a note)

    Returns ``(connections, unusable)``.
    """
    unusable = sorted(i for i, m in graph.items() if not m)

    # connected components over the "mates with" relation (ignoring unusable)
    remaining = set(i for i in graph if graph[i])
    comps = []
    while remaining:
        seed = next(iter(remaining))
        stack, comp = [seed], set()
        while stack:
            x = stack.pop()
            if x in comp:
                continue
            comp.add(x)
            for y in graph[x]:
                if y in remaining and y not in comp:
                    stack.append(y)
        remaining -= comp
        comps.append(sorted(comp))

    connections = []
    for comp in comps:
        if len(comp) == 1:
            connections.append((comp[0], comp[0]))          # self-mate
        elif len(comp) == 2:
            connections.append((comp[0], comp[1]))          # congruent pair
        else:
            for k in range(0, len(comp) - 1, 2):            # generic: pair up
                connections.append((comp[k], comp[k + 1]))
    return connections, unusable

File: src/core/mating.ts
/**
 * mating.ts — which faces can be glued face-to-face on an IDENTICAL copy, and
 * the rigid transform that performs the join. Port of mating.py.
 *
 * Congruence is necessary but NOT sufficient: a flush joint needs the two faces
 * to meet with outward normals OPPOSED, i.e. one polygon must coincide with the
 * mirror image of the other. So a chiral face related to its partner only by a
 * rotation cannot mate (the mirror-shaped face does not exist on an identical
 * copy). We detect this numerically: reverse one face's vertex order (opposing
 * normals flip orientation), search cyclic correspondences with a proper Kabsch
 * fit, and accept a mate only when rmsd ≈ 0 AND the normals end up anti-parallel.
 */
import { dot } from './vec3.ts';
import type { Vec3 } from './vec3.ts';
import { kabsch, matVec, type Rigid } from './linalg.ts';
import type { Face } from './geometry.ts';

export interface Mate extends Rigid {
  nd: number; // normal·normal after the fit; a clean mate has nd ≈ −1
}

// np.roll(arr, s): result[i] = arr[(i − s) mod n]
const roll = <T>(arr: readonly T[], s: number): T[] => {
  const n = arr.length;
  return arr.map((_, i) => arr[((i - s) % n + n) % n]);
};

/**
 * Proper rigid motion laying `faceJ` flush onto `faceI` (centroids coincide,
 * normals opposed, outlines aligned). Returns the best correspondence, or null
 * if vertex counts differ.
 */
export function mateTransform(faceJ: Face, faceI: Face): Mate | null {
  const A0 = faceJ.pts.slice().reverse(); // reverse: opposing normals flip orientation
  const B = faceI.pts;
  if (A0.length !== B.length) return null;
  let best: Mate | null = null;
  for (let s = 0; s < B.length; s++) {
    const fit = kabsch(roll(A0, s), B);
    if (best === null || fit.rmsd < best.rmsd) {
      const nd = dot(matVec(fit.R, faceJ.normal) as Vec3, faceI.normal);
      best = { ...fit, nd };
    }
  }
  return best;
}

export type MateGraph = Map<number, number[]>;

/** For every face, the sorted list of faces it can be glued flush against. */
export function flushMateGraph(faces: Face[], rmsdTol = 1e-2): MateGraph {
  const byIdx = new Map(faces.map((f) => [f.idx, f]));
  const graph: MateGraph = new Map();
  for (const i of faces) {
    const mates: number[] = [];
    for (const j of faces) {
      const res = mateTransform(byIdx.get(j.idx)!, byIdx.get(i.idx)!);
      if (res && res.rmsd < rmsdTol && res.nd < -0.99) mates.push(j.idx);
    }
    graph.set(
      i.idx,
      mates.sort((a, b) => a - b),
    );
  }
  return graph;
}

export interface Connections {
  connections: [number, number][];
  unusable: number[];
}

/**
 * Turn the mate graph into connections + unusable faces:
 *   • empty mate list      → unusable
 *   • component of size 1   → self-connection (i, i)
 *   • congruent pair {a, b} → connection (a, b)
 *   • larger components chained pairwise
 * Port of derive_connections.
 */
export function deriveConnections(graph: MateGraph): Connections {
  const unusable = [...graph.entries()]
    .filter(([, m]) => m.length === 0)
    .map(([i]) => i)
    .sort((a, b) => a - b);

  const remaining = new Set([...graph.entries()].filter(([, m]) => m.length > 0).map(([i]) => i));
  const comps: number[][] = [];
  while (remaining.size > 0) {
    const seed = remaining.values().next().value as number;
    const stack = [seed];
    const comp = new Set<number>();
    while (stack.length > 0) {
      const x = stack.pop()!;
      if (comp.has(x)) continue;
      comp.add(x);
      for (const y of graph.get(x)!) if (remaining.has(y) && !comp.has(y)) stack.push(y);
    }
    for (const x of comp) remaining.delete(x);
    comps.push([...comp].sort((a, b) => a - b));
  }

  const connections: [number, number][] = [];
  for (const comp of comps) {
    if (comp.length === 1) connections.push([comp[0], comp[0]]);
    else if (comp.length === 2) connections.push([comp[0], comp[1]]);
    else for (let k = 0; k + 1 < comp.length; k += 2) connections.push([comp[k], comp[k + 1]]);
  }
  return { connections, unusable };
}
